File: PythonCodes/Fig05_01.py
#%%
"""
Created on Thu Jan 11 2019
Paths for the Poisson process and the compensated Poisson process
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt

def GeneratePathsPoisson(NoOfPaths,NoOfSteps,T,xiP):    

    # Create empty matrices for Poisson and compensated Poisson process

    X = np.zeros([NoOfPaths, NoOfSteps+1])
    Xc = np.zeros([NoOfPaths, NoOfSteps+1])
    time = np.zeros([NoOfSteps+1])
                
    dt = T / float(NoOfSteps)
    
    Z = np.random.poisson(xiP*dt,[NoOfPaths,NoOfSteps])
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        X[:,i+1]  = X[:,i] + Z[:,i]
        Xc[:,i+1] = Xc[:,i] -xiP*dt + Z[:,i]
        time[i+1] = time[i] +dt
        
    paths = {"time":time,"X":X,"Xcomp":Xc}
    return paths

def mainCalculation():
    NoOfPaths = 25
    NoOfSteps = 500
    T = 30
    xiP= 1
        
    Paths = GeneratePathsPoisson(NoOfPaths,NoOfSteps,T,xiP)
    timeGrid = Paths["time"]
    X = Paths["X"]
    Xc = Paths["Xcomp"]
       
    plt.figure(1)
    plt.plot(timeGrid, np.transpose(X),'-b')   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("X(t)")
    
    plt.figure(2)
    plt.plot(timeGrid, np.transpose(Xc),'-b')   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("X(t)")
                     
mainCalculation()

File: MatlabCodes/Chapter 05/Fig05_01.m
function PoissonProcess_paths
clc;close all;

% Monte Carlo settings

NoOfPaths = 25;
NoOfSteps = 500;
T         = 30;
dt        = T/NoOfSteps;
time      = 0:dt:T;

% Poisson process settings

xiP       = 1;

% Empty matrices for the Poisson paths

X  = zeros(NoOfPaths,NoOfSteps);
Xc =zeros(NoOfPaths,NoOfSteps);

% Random noise

Z = random('poisson',xiP*dt,[NoOfPaths,NoOfSteps]);

% Creation of the paths

for i=1:NoOfSteps
    X(:,i+1)  = X(:,i) + Z(:,i);
    Xc(:,i+1) = Xc(:,i) - xiP * dt + Z(:,i);
end
figure(1)
plot(time,X,'color',[0 0.45 0.74],'linewidth',1.2)
xlabel('time')
ylabel('$$X_\mathcal{P}(t)$$','interpreter','latex')
grid on

figure(2)
plot(time,Xc,'color',[0 0.45 0.74],'linewidth',1.2)
xlabel('time')
ylabel('$$X_\mathcal{P}(t)-\xi_pdt$$','interpreter','latex')
grid on
